R/sidebar_layout.R

```diff
@@ -0,0 +1,208 @@
+# ------------------------------------------------------------------------------
+# Sidebar layout composed of main and sidebar panels
+# ------------------------------------------------------------------------------
+#'
+#' Creates row wrapping specified elements
+#' @param arg Element to be wrapped in row container
+#' @return Row wrapping specified elements
+
+get_row <- function(arg) {
+  class <- "row"
+  style <- "padding: 20px;"
+  HTML(glue::glue("<div class='{class}' style='{style}'>{arg}</div>"))
+}
+
+#' Creates div containing elements in order or grid container for specific panels
+#'
+#' @param grid_list List to create grid from
+#' @param ... Container's children elements
+#' @return Div containing elements or grid container for specific panel
+
+panel <- function(grid_list, ...) {
+  args <- list(...)
+  if(is.null(grid_list)) {
+    div(lapply(args, get_row))
+  } else {
+    grid(grid_list$layout, grid_list$container_style, grid_list$area_styles, ...)
+  }
+}
+
+#' Creates div containing elements in order or grid container for sidebar panel
+#'
+#' @param grid_list List to create grid from
+#' @param ... Container's children elements
+#' @export
+
+sidebar_panel <- function(grid_list, ...) {
+  panel(grid_list, ...)
```

Comment:
so tbh I don't seem to understand why we need `sidebar_panel` and `main_panel` here as they both call panel without any specific for each elements. is it just shiny-like wrapper?

Reply:
sidebar_panel and main_panel are the only two children elements of sidebar_layout grid container and they need to have some grid-area names in order to make grid function work. So sidebar_panel and main_panel are both shiny-like wrappers and grid-area names. I didn't know how to optimize the code more assuming that right now both of them call the same panel() function, without any specific additional calls for each one respectively.

Reply:
I used grid_list and ... separately, instead of using just grid() function in case user doesn't want to specify grid areas and wants to specify only children elements of grid (so the're displayed one under each other, as it normally occurs in container with display: grid applied in css, but with no grid-areas specified). Currently grid() function requires to specify grid-areas as children elements, so I wanted temporarily to allow user an option to just display elements without grid-areas applied to them